openhtf/__init__.py

```diff
@@ -45,6 +46,137 @@ class InvalidTestPhaseError(Exception):
   """Raised when an invalid method is decorated."""
 
 
+class Test(object):
+  """An object that represents an OpenHTF test.
+
+  Example:
+
+    def PhaseOne(test):
+      # Integrate more widgets
+
+    def PhaseTwo(test):
+      # Analyze widget integration status
+
+    Test(PhaseOne, PhaseTwo).Execute()
+  """
+
+  def __init__(self, *phases, **metadata):
+    code_info = test_record.CodeInfo.ForModuleFromStack(levels_up=2)
+    self._test_info = TestInfo(phases, metadata=metadata, code_info=code_info)
+    self._output_callbacks = []
+    self.loop = False
+
+  def AddOutputCallback(self, callback):
+    """Add the given function as an output module to this test."""
+    self._output_callbacks.append(callback)
+
+  def OutputTestRecord(self, record):
+    """Feed the record of this test to all output modules."""
+    record.metadata.update(self._test_info.metadata)
+    for output_cb in self._output_callbacks:
+      output_cb(record)
+
+  # TODO(fahhem): Cleanup accesses to these attributes and remove these proxies.
+  @property
+  def plug_type_map(self):
+    return self._test_info.plug_type_map
+
+  @property
+  def phases(self):
+    return self._test_info.phases
+
+  @property
+  def code_info(self):
+    return self._test_info.code_info
+
+  @property
+  def metadata(self):
+    return self._test_info.metadata
+
+  def Execute(self, loop=None,
+              test_start=triggers.AutoStart, test_stop=triggers.AutoStop):
+    """Starts the framework and executes the given test.
+    Args:
+      test_start: Trigger for starting the test, defaults to AutoStart with a
+          dummy serial number.
+      test_stop: Trigger for when the test is over, defaults to AutoStop to
+          immediately stop after the phase.
+      output_callbacks: List of callbacks to be called with the results
+          output from this test.
+    """
+    SetupFramework()
+
+    if loop is not None:
+      self.loop = loop
+
+    _LOG.info('Executing test: %s', self.code_info.name)
+    executor = exe.TestExecutor(conf.Config(), self, test_start, test_stop)
+    server = http_api.Server(executor)
+    StopOnSigInt([server.Stop, executor.Stop])
+    server.Start()
+    executor.Start()
+    executor.Wait()
+    server.Stop()
+
+
+class TestInfo(mutablerecords.Record(
+    'Test', ['phases', 'code_info'], {'metadata': dict})):
+  """An object that represents the reusable portions of an OpenHTF test.
+
+  This object encapsulates the static test information that is set once and used
```

Comment:
If it's set once, does it need to be a `Record` instead of a `namedtuple`?  It looks like the only place it gets instantiated, `metadata` is always provided, so neither the mutability nor the the default value mechanism of the `Record` are used, it looks like?